spec/unit/resource_spec.rb

```diff
@@ -336,6 +336,109 @@ class ResourceTestHarness < Chef::Resource
     end
   end
 
+  describe "self.resource_name" do
+    context "When resource_name is not set" do
+      it "and there is a provides line, it is used for resource_name" do
+        c = Class.new(Chef::Resource) do
+          provides :self_resource_name_test_1
+        end
+
+        r = c.new('hi')
+        r.declared_type = :d
+        expect(c.resource_name).to eq :self_resource_name_test_1
+        expect(r.resource_name).to eq :self_resource_name_test_1
+        expect(r.declared_type).to eq :d
+      end
+      it "and multiple provides lines, the first line is used for resource_name" do
+        c = Class.new(Chef::Resource) do
+          provides :self_resource_name_test_2
+          provides :self_resource_name_test_3
+        end
+
+        r = c.new('hi')
+        r.declared_type = :d
+        expect(c.resource_name).to eq :self_resource_name_test_2
+        expect(r.resource_name).to eq :self_resource_name_test_2
+        expect(r.declared_type).to eq :d
+      end
+      it "and there are no provides lines, resource_name is nil" do
```

Comment:
If it's not valid to have a `nil` resource name, is this testing that we successfully get ourselves to an invalid state without an error? It might be worth a comment, at least.